### PR TITLE
Add Pixtral/Ministral3 VLM support: dynamic vision, FP8 dequant, PatchMerger fix, MHA

### LINKED ISSUE #151: Pixtral vision model: ORT fused RotaryEmbedding op produces wrong features for 2D spatial RoPE
## Summary

The Pixtral vision encoder exported by mobius produces incorrect features due to ORT's fused `RotaryEmbedding` contrib op not supporting Pixtral's 2D spatial RoPE. The model's attention layers effectively produce near-identical outputs for different inputs (cosine similarity 0.984 between random inputs), making the VLM unable to understand images.

## Root Cause

Pixtral uses **2D spatial RoPE** (separate frequency bands for height and width coordinates, concatenated), which differs from the standard 1D sequential RoPE used in text models. When mobius exports the vision model, ORT's graph optimizer fuses the attention + RoPE ops into:
- `com.microsoft.Attention` (fused attention op)
- `com.microsoft.RotaryEmbedding` (fused RoPE op)

The fused `RotaryEmbedding` contrib op was designed for **1D text RoPE** and does not correctly handle Pixtral's 2D cos/sin position embeddings. This causes the attention mechanism to produce effectively position-agnostic features.

## Evidence

**Broken model (fused ops, 405 nodes):**
- Cosine similarity vs HuggingFace PyTorch: **0.014** (essentially random)
- Feature norm: 358.65 (6x higher than HF's 58.61)
- Random input independence: cosine **0.984** (model doesn't differentiate inputs)
- Uses: `Attention` (24) + `RotaryEmbedding` (48) fused ops

**Working model (unfused ops, 1032 nodes, via torch.export):**
- Cosine similarity vs HuggingFace PyTorch: **0.9994** (near-perfect match)
- Feature norm: 58.53 (matches HF's 58.61)
- Random input independence: cosine **0.347** (model correctly differentiates inputs)
- Uses: `MultiHeadAttention` (24) + explicit `Cos/Sin/Mul/Slice/Neg` ops for RoPE

**Key difference:** The working model implements RoPE with standard ONNX ops, avoiding the fused `RotaryEmbedding` contrib op entirely.

## Additional Details

- Graph structure is correct: all 24 transformer layers are wired, 96 attention weight_t initializers are connected
- Weights are correct: cosine similarity 1.0 between HF and ONNX weights, max diff 0.0
- Attention op attributes are correct: q_num_heads=16, kv_num_heads=16, scale=0.125
- RotaryEmbedding attributes: num_heads=16, rotary_embedding_dim=0 (0 means "full dim")
- The issue is specifically in the **runtime execution** of the fused RotaryEmbedding with 2D cos/sin inputs

## Proposed Fix

**Option A (recommended):** Change `PixtralAttention` (or `apply_rotary_pos_emb`) to implement RoPE with standard ONNX ops (`Cos`, `Sin`, `Mul`, `Slice`, `Neg`, `Concat`) instead of relying on ORT's fused `RotaryEmbedding` contrib op. This matches the working torch.export approach.

**Option B:** Use the ONNX standard `RotaryEmbedding` op (opset 23) instead of the ORT contrib op. Mobius LLM models already use the opset 23 standard op — the vision path should too.

**Option C:** Investigate and fix ORT's fused `RotaryEmbedding` contrib op to support 2D spatial cos/sin inputs. This is an upstream ORT change and would benefit all 2D RoPE models.

## Testing

A `test_vision_features_parity` test was added to `tests/integration_test.py` (TestMistral3VL3Model) that checks:
1. Cosine similarity > 0.99 vs HuggingFace reference
2. Norm ratio in [0.9, 1.1]
3. Random input independence < 0.9

This test is gated behind HF model access. **Recommendation:** Add a lightweight synthetic parity test that builds a small random-weight Pixtral vision model, exports both fused and unfused versions, and compares outputs. This would run in CI without HF access.

## Affected Models

- `mistralai/Ministral-3-3B-Instruct-2512` (Pixtral vision encoder)
- Any future model using Pixtral's 2D spatial RoPE in the vision encoder

## References

- PR #130 (PixtralVLTask addition)
- onnxruntime-genai PR #2077 (Mistral3 VLM support)
- olive-recipes PR #352 (Ministral VLM recipe)
- Reference working model: `~/Ministral-3-3B-Instruct-ONNX` (torch.export based)

### PR DESCRIPTION
**Pull request by titaiwangms**
## Summary

Adds Pixtral/Ministral3 VLM support with 3-model split (vision/embedding/decoder) for onnxruntime-genai.

### Vision encoder (`_pixtral_vision.py`)
- `PixtralVisionTower`: Conv2d patch embedding → RMSNorm → 24-layer transformer with 2D RoPE
- `PixtralRoPE2D`: Precomputed 2D rotary position embeddings over spatial grid
- `PixtralAttention`: Bidirectional MHA with 2D RoPE (`com.microsoft.MultiHeadAttention`; `op.Attention` fallback for onnx-standard EP)
- `Mistral3PatchMerger`: Spatial 2×2 patch merging matching HF `F.unfold` dim-major ordering
- `Mistral3MultiModalProjector`: norm → merge → GELU MLP projection

### PatchMerger fix (cosine sim 0.007 → 0.999973)
The original transpose permutation `[0,1,3,2,4,5]` produced patch-major ordering, but HF `F.unfold` produces dim-major `[D, ms_h, ms_w]`. The learned `merging_layer` weights expect HF ordering. Fixed to `[0,1,3,5,2,4]`.

### FP8 weight dequantization (`_weight_loading.py`)
- Detects FP8 weights, dequantizes via `fp8.to(bf16) * scale`
- Always returns new dict (non-mutating, addresses review comment)
- Docstring updated to list all removed suffixes

### Attention scale optimization (`_attention.py`)
- Moved `attn_scale` Mul before RoPE (scaling commutes with rotation) to preserve `RotaryEmbedding → Attention` pattern for GQA rewrite rule matching

### Task and infrastructure
- `PixtralVLTask`: Dynamic H×W vision input with Squeeze for rank-2 output
- Registered as `pixtral-vl` task in `_vision_language_3model.py`
- Olive-recipe `optimize.py` integration verified

## Testing
- Cosine sim vs HuggingFace PyTorch: **0.999973**
- E2E `model-mm.py` with fish.jpg and challenge.jpg: correct descriptions
- Olive-recipe pipeline export + E2E verified
- `test_patch_merger_matches_hf_unfold_ordering`: regression test against HF F.unfold
- All 1261 unit tests pass, lintrunner clean

## Files Changed
- `src/mobius/components/_pixtral_vision.py` — PatchMerger fix + MHA
- `src/mobius/components/_pixtral_vision_test.py` — F.unfold ordering test
- `src/mobius/tasks/_vision_language_3model.py` — Squeeze vision output
- `src/mobius/_weight_loading.py` — Docstring + mutation fix
- `src/mobius/components/_attention.py` — attn_scale before RoPE

### CODE REVIEW

File: src/mobius/_registry.py [hunk 1]
```diff
@@ -329,328 +328,380 @@ def _detect_fallback_registration(hf_config) -> ModelRegistration | None:
...
+        "got_ocr2",
+        "janus",
+        "ovis2",
+        "vipllava",
+    ):
+        reg.register(name, LLaVAModel, task="vision-language")
+
+    # Pixtral-based models use dynamic-resolution vision encoder
+    for name in ("mistral3", "pixtral"):
+        reg.register(name, LLaVAModel, task="pixtral-vl")
```
Comment (justinchuby) [inline src/mobius/_registry.py]: I believe this was updated in main. Can you rebase?

File: src/mobius/_registry.py [hunk 2]
```diff
@@ -498,8 +498,7 @@ def _detect_fallback_registration(hf_config) -> ModelRegistration | None:
     "qwen3_vl_text": ModelRegistration(Qwen3VLTextModel),
     "smolvlm": ModelRegistration(LLaVAModel, task="vision-language"),
     "video_llava": ModelRegistration(LLaVAModel, task="vision-language"),
-    "vipllava": ModelRegistration(LLaVAModel, task="vision-language"),
-    # --- Speech ---
+    "vipllava": ModelRegistration(LLaVAModel, task="vision-language"),    # --- Speech ---
```
Comment (justinchuby) [inline src/mobius/_registry.py]:
```suggestion
    "vipllava": ModelRegistration(LLaVAModel, task="vision-language"),
    # --- Speech ---
```
nit

File: src/mobius/_weight_loading.py [hunk 3]
```diff
@@ -171,4 +221,6 @@ def _download_weights(model_id: str) -> dict[str, torch.Tensor]:
     state_dict: dict[str, torch.Tensor] = {}
     for path in tqdm.tqdm(paths, desc="Loading weights"):
         state_dict.update(safetensors.torch.load_file(path))
+
+    state_dict = _dequantize_fp8_weights(state_dict)
```
Comment (justinchuby) [inline src/mobius/_weight_loading.py:227]: Is it possible to do it where we create ir Tensors, so that we can use LazyTensor?

File: testdata/cases/vision-language/ministral-3-3b.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  images:
+    - "pipeline-cat-chonk.jpeg"
+
+level: "L4+L5"
+
+generation:
+  max_new_tokens: 30
+  do_sample: false
+
+skip_reason: "Gated repo (mistralai/Ministral-3-3B-Instruct-2512 requires authentication)."
```
Comment (justinchuby) [inline testdata/cases/vision-language/ministral-3-3b.yaml:18]: Is it possible to generate the json data with your hf login?

**Comment by justinchuby** [comment]
Only comments are LazyTensor + Referece json files, which can also be a follow up if needed

**Review by justinchuby — APPROVED** [review]

File: src/mobius/models/llava.py [hunk 5]
```diff
@@ -129,7 +129,20 @@ def forward(self, op: builder.OpBuilder, input_ids: ir.Value, image_features: ir
         indices = op.Sub(cumsum, op.Constant(value_int=1))
         indices = op.Clip(indices, op.Constant(value_int=0))
 
-        gathered = op.Gather(image_features, indices, axis=0)
+        # Pad image_features with one zero row so Gather is valid even when
```
Comment (justinchuby) [inline src/mobius/models/llava.py:132]: I came across the same thing in gemma4. Maybe work extracting this logic once we merge both

File: src/mobius/_weight_loading.py (hunk 3 above)
Comment (titaiwangms) [reply src/mobius/_weight_loading.py:227]: FP8 dequantization involves multiplication (`fp8.to(bf16) * scale`), not just a dtype cast, so it can't be deferred via `ir.LazyTensor` directly. The downstream `_assign_weight()` already uses `LazyTensor` for dtype casts at serialization time. Restructuring to defer the multiplication would require carrying scale metadata through the state dict pipeline — worth considering as a follow-up.

File: src/mobius/tasks/__init__.py [hunk 6]
```diff
@@ -79,6 +80,7 @@
 from mobius.tasks._vision_language_3model import (
     HybridQwenVLTask,
     MllamaVisionLanguageTask,
+    PixtralVLTask,
     QwenVLTask,
     VisionLanguageTask,
 )
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/mobius/tasks/__init__.py:94]: PR description updated to reflect the actual implementation location (`_vision_language_3model.py`, registered as `pixtral-vl` task). No separate `pixtral_vl.py` file exists.

File: src/mobius/tasks/_vision_language_3model.py [hunk 7]
```diff
@@ -356,6 +356,52 @@ def _build_decoder(
...
 
 
+class PixtralVLTask(VisionLanguageTask):
+    """Vision-language task with dynamic-resolution Pixtral vision encoder.
+
+    Pixtral's internal computations (patch embedding, 2D RoPE, spatial merge)
+    are fully dynamic — grid_h and grid_w are derived from ``op.Shape()`` at
+    runtime.  This subclass replaces the static ``image_size`` input shape
+    with symbolic ``height`` / ``width`` dimensions so the exported ONNX
+    model accepts variable-resolution images.
```
[the remark this answers is not in the document]
Comment (titaiwangms) [reply src/mobius/tasks/_vision_language_3model.py:197]: Fixed — PR description updated to match the actual code structure.

File: testdata/cases/vision-language/ministral-3-3b.yaml (hunk 4 above)
Comment (titaiwangms) [reply testdata/cases/vision-language/ministral-3-3b.yaml:18]: The model is gated (requires mistralai authentication). The `skip_reason` field documents this. We could add HF auth token support to CI as a follow-up to enable integration tests for gated repos.

File: src/mobius/models/llava.py (hunk 5 above)
Comment (titaiwangms) [reply src/mobius/models/llava.py:132]: Agreed — the image embedding fusion logic (masked_scatter pattern) is shared with Gemma4. Worth extracting into a shared utility once both PRs land.

File: testdata/cases/vision-language/ministral-3-3b.yaml (hunk 4 above)
Comment (titaiwangms) [reply testdata/cases/vision-language/ministral-3-3b.yaml:18]: Follow -up task created.

File: src/mobius/_weight_loading.py (hunk 3 above)
Comment (titaiwangms) [reply src/mobius/_weight_loading.py:227]: @justinchuby , this should be covered.

Comment (titaiwangms) [reply src/mobius/_weight_loading.py:227]: Added it to follow-up to see if there is a way.

**Comment by justinchuby** [comment]
@titaiwangms if you have a Olive-recipe e2e, please feel free to add it to examples/olive/ as a demo. Thanks!

**Comment by justinchuby** [comment]
@titaiwangms CI failing: https://github.com/onnxruntime/mobius/actions/runs/24267296245/job/70864946872?pr=157

is there a way to run the new ort test without saving model to disk?

**Comment by titaiwangms** [comment]
> @titaiwangms CI failing: https://github.com/onnxruntime/mobius/actions/runs/24267296245/job/70864946872?pr=157
> 
> is there a way to run the new ort test without saving model to disk?

let me fix it.
